Aplib.Core/Action.cs

```diff
@@ -0,0 +1,36 @@
+﻿using System;
+
+namespace Aplib.Core
+{
+    /// <summary>
+    /// Describes an action that can be executed and guarded.
+    /// </summary>
+    /// <typeparam name="TQuery">The type of the query of the action</typeparam>
+    public class Action<TQuery>
+    {
+        private TQuery? _storedResult;
+
+        /// <summary>
+        /// The effect of the action.
+        /// </summary>
+        public System.Action<TQuery>? Effect { private get; set; }
+
+        public bool Actionable => Query!.Invoke() is not (false or null);
```

Comment:
- Why should Query never be `null`? I made a mistake in my previous comment; I said the Query being `null` should cause the Action to be unactionable, but I think it should be the opposite. Query should be used to guard the action. If the user doesn't want the action to be guarded, it should be possible to simply not specify a Query (then the effect would always be applied). What would be the benefit of not allowing Query to be `null`?
- I agree that giving a warning would not be ideal. Errors, however, have the same flaw in my opinion. The constructor **does** solve my issue, though, since this makes it impossible to implement such unwanted states.
- I think property initializers are well-suited for **optional** properties. Instead, constructors have the exact purpose we want here.
- Even if it was a compile-time error, it might cause confusion amongst developers, as it is not how C# is designed.